### PR TITLE
API-1391: Keep events even if the events API limit is reached

### PR DESCRIPTION
**Pull request by pierrejolly**
Today, the webhook requests limit has a role in limiting the number of events API sent but also in purging messages as the events we cannot send are consumed. 

The webhook requests limit role is to stop sending webhook requests and wait for the next slot to send them again. The hard limit has been mainly created to limit the costs for scalable SaaS environments and define what a customer can expect from this Events API.

This will help in losing fewer messages and be compliant with more customers

### CODE REVIEW

File: src/Akeneo/Connectivity/Connection/back/Infrastructure/EventSubscriber/MessengerEventSubscriber.php [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+use Psr\Log\LoggerInterface;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\Messenger\Event\WorkerRunningEvent;
+
+/**
+ * @author Pierre Jolly <pierre.jolly@akeneo.com>
+ * @copyright 2021 Akeneo SAS (http://www.akeneo.com)
+ * @license http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+final class MessengerEventSubscriber implements EventSubscriberInterface
```
Comment (pchasle) [inline src/Akeneo/Connectivity/Connection/back/Infrastructure/EventSubscriber/MessengerEventSubscriber.php:19]:
```suggestion
final class EventsApiRequestsLimitEventSubscriber implements EventSubscriberInterface
```

File: src/Akeneo/Connectivity/Connection/back/Infrastructure/EventSubscriber/MessengerEventSubscriber.php [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    public function checkWebhookRequestLimit(): void
+    {
+        $delayUntilNextRequest = $this->getDelayUntilNextRequest->execute(
+            new \DateTimeImmutable('now', new \DateTimeZone('UTC')),
+            $this->webhookRequestsLimit
+        );
+
+        if (0 <= $delayUntilNextRequest) {
```
Comment (pchasle) [inline src/Akeneo/Connectivity/Connection/back/Infrastructure/EventSubscriber/MessengerEventSubscriber.php:52]:
```suggestion
        if ($delayUntilNextRequest > 0) {
```
Easier to read this way.

And I think you should not log anything or sleep if the result is 0 ?!

File: src/Akeneo/Connectivity/Connection/back/Infrastructure/Persistence/Dbal/Query/DbalSelectEventsApiRequestCountWithinLastHourQuery.php [hunk 3]
```diff
@@ -21,16 +21,17 @@ public function __construct(Connection $dbalConnection)
...
+    public function execute(\DateTimeImmutable $eventDateTime): array
     {
         $oneHourAgoEventDateTime = $eventDateTime->modify('-1 hour');
         $sql = <<<SQL
- SELECT sum(event_count)	
- FROM akeneo_connectivity_connection_events_api_request_count 	
- WHERE updated BETWEEN :from_datetime  AND :to_datetime	
+ SELECT updated, event_count
+ FROM akeneo_connectivity_connection_events_api_request_count
+ WHERE updated BETWEEN :from_datetime  AND :to_datetime
```
Comment (pchasle) [inline src/Akeneo/Connectivity/Connection/back/Infrastructure/Persistence/Dbal/Query/DbalSelectEventsApiRequestCountWithinLastHourQuery.php:30]:
```suggestion
 WHERE updated BETWEEN :from_datetime AND :to_datetime
```

File: src/Akeneo/Connectivity/Connection/back/Infrastructure/Webhook/Service/Sleep.php [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+namespace Akeneo\Connectivity\Connection\Infrastructure\Webhook\Service;
+
+/**
+ * @author Pierre Jolly <pierre.jolly@akeneo.com>
+ * @copyright 2021 Akeneo SAS (http://www.akeneo.com)
+ * @license http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class Sleep
+{
+    public function execute(int $second): void
```
Comment (pchasle) [inline src/Akeneo/Connectivity/Connection/back/Infrastructure/Webhook/Service/Sleep.php:14]:
```suggestion
    public function sleep(int $second): void
```

File: src/Akeneo/Connectivity/Connection/back/Infrastructure/Webhook/Service/GetDelayUntilNextRequest.php [hunk 5]
```diff
@@ -0,0 +1,39 @@
+<?php
+
```
Comment (willy-ahva) [inline src/Akeneo/Connectivity/Connection/back/Infrastructure/Webhook/Service/GetDelayUntilNextRequest.php:2]: declare strict type

File: src/Akeneo/Connectivity/Connection/back/tests/Integration/Persistence/Dbal/Query/SelectEventsApiRequestCountWithinLastHourQueryIntegration.php [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+        $this->eventsApiRequestCountLoader = $this->get(
+            'akeneo_connectivity.connection.fixtures.events_api_request_count_loader'
+        );
+        $this->eventsApiRequestCountWithinLastHour = $this->get(
+            'akeneo_connectivity.connection.persistence.query.select_events_api_request_count_within_last_hour_query'
+        );
+    }
+
+    public function test1()
```
Comment (willy-ahva) [inline src/Akeneo/Connectivity/Connection/back/tests/Integration/Persistence/Dbal/Query/SelectEventsApiRequestCountWithinLastHourQueryIntegration.php:33]: rename

File: src/Akeneo/Connectivity/Connection/back/tests/Integration/Persistence/Dbal/Query/SelectEventsApiRequestCountWithinLastHourQueryIntegration.php [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+use Akeneo\Connectivity\Connection\Domain\Audit\Persistence\Query\SelectEventsApiRequestCountWithinLastHourQuery;
+use Akeneo\Test\Integration\TestCase;
+use PHPUnit\Framework\Assert;
+
+/**
+ * @author    Thomas Galvaing <thomas.galvaing@akeneo.com>
+ * @copyright 2020 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class SelectEventsApiRequestCountWithinLastHourQueryIntegration extends TestCase
```
Comment (willy-ahva) [inline src/Akeneo/Connectivity/Connection/back/tests/Integration/Persistence/Dbal/Query/SelectEventsApiRequestCountWithinLastHourQueryIntegration.php:16]: you must test that if there is no result it returns an array

**Review by willy-ahva — APPROVED** [review]

File: src/Akeneo/Connectivity/Connection/back/Infrastructure/Webhook/Service/GetDelayUntilNextRequest.php [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+{
+    private SelectEventsApiRequestCountWithinLastHourQuery $selectEventsApiRequestCountWithinLastHourQuery;
+
+    public function __construct(
+        SelectEventsApiRequestCountWithinLastHourQuery $selectEventsApiRequestCountWithinLastHourQuery
+    ) {
+        $this->selectEventsApiRequestCountWithinLastHourQuery = $selectEventsApiRequestCountWithinLastHourQuery;
+    }
+
+    public function execute(\DateTimeImmutable $dateTime, $limit): int
```
Comment (pchasle) [inline src/Akeneo/Connectivity/Connection/back/Infrastructure/Webhook/Service/GetDelayUntilNextRequest.php:22]:
```suggestion
    public function execute(\DateTimeImmutable $dateTime, int $limit): int
```

Comment (pchasle) [inline src/Akeneo/Connectivity/Connection/back/Infrastructure/Webhook/Service/GetDelayUntilNextRequest.php:22]: We need to add an explanation on what is going on here.
It's hard to follow the fact we are fetching the count for each minutes ordered from latest to oldest then we are adding the count together until the limit is reached to finally calculate the number of seconds between the current time and the time where the limit was reached + adding the remaining time in seconds to "complete" the 1 hour quota. :scream:

Comment (pierrejolly) [reply src/Akeneo/Connectivity/Connection/back/Infrastructure/Webhook/Service/GetDelayUntilNextRequest.php:22]: Yes indeed, I'll add a comment

File: src/Akeneo/Connectivity/Connection/back/tests/Unit/spec/Infrastructure/Webhook/Service/GetDelayUntilNextRequestSpec.php [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+                    'updated' => '2021-01-08 10:32:30',
+                ],
+                [
+                    'event_count' => 90,
+                    'updated' => '2021-01-08 10:11:30',
+                ],
+            ]);
+
+        $this->execute(new \DateTimeImmutable('2021-01-08 11:02:10', new \DateTimeZone('UTC')), 100)
+            ->shouldReturn(560);
```
Comment (pchasle) [inline src/Akeneo/Connectivity/Connection/back/tests/Unit/spec/Infrastructure/Webhook/Service/GetDelayUntilNextRequestSpec.php:49]:
```suggestion
                    'event_count' => 20,
                    'updated' => '2021-01-08 10:32:30',
                ],
                [
                    'event_count' => 90,
                    'updated' => '2021-01-08 10:11:30', // Limit will be reached here, minute xx:11:30 = 690 seconds.
                ],
            ]);

        $this->execute(new \DateTimeImmutable('2021-01-08 11:02:10', new \DateTimeZone('UTC')), 100)
            ->shouldReturn(560); // Current time minute is xx:02:10 = 130 seconds, so 690 - 130 = 560 seconds
```

**Review by pchasle — APPROVED** [review]
